abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+`service_name_size`).
+
+Returns `NotAllowed` when requests to given service are not allowed.
+
+Returns `InvalidMemoryAccess` when `grpc_service_data`, `grpc_service_size`,
+`service_name_data`, `service_name_size`, `method_data`, `method_size`,
+`initial_metadata_map_data`, `initial_metadata_map_size` and/or
+`return_callout_id` point to invalid memory address.
+
+#### `proxy_send_grpc_stream_message`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+  - `i32 (iovec_array_t) grpc_message_data`
+  - `i32 (size_t) grpc_message_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Send gRPC message (`grpc_message_data`, `grpc_message_size`) on the existing
+gRPC stream (`callout_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the gRPC call (`callout_id`) doesn't exist.
+
+Returns `InvalidMemoryAccess` when `grpc_message_data` and/or
+`grpc_message_size` point to invalid memory address.
+
+#### `proxy_cancel_grpc_call`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Cancel outstanding gRPC request or stream (`callout_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the gRPC call (`callout_id`) doesn't exist.
+
+#### `proxy_close_grpc_call`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Cancel outstanding gRPC request or close existing gRPC stream (`callout_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the gRPC call (`callout_id`) doesn't exist.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_grpc_call_response_header_metadata`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) num_headers`
+* returns:
+  - none
+
+Called when header metadata in the response to the gRPC call (`callout_id`) is
+received.
+
+Header metadata (`num_headers` entries) can be retrieved using
+`proxy_get_map_values`.
+
+#### `proxy_on_grpc_call_response_message`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) message_size`
+* returns:
+  - none
+
+Called when the response message (of size `message_size`) to the gRPC call
+(`callout_id`) is received.
+
+Content of the gRPC message (of size `data_size`) can be retrieved using
+`proxy_get_buffer`.
+
+#### `proxy_on_grpc_call_response_trailer_metadata`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+  - `i32 (size_t) num_trailers`
+* returns:
+  - none
+
+Called when trailer metadata in the response to the gRPC call (`callout_id`) is
+received.
+
+Trailer metadata (`num_trailers` entries) can be retrieved using
+`proxy_get_map_values`.
+
+#### `proxy_on_grpc_call_close`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+  - `i32 (uint32_t) status_code`
+* returns:
+  - none
+
+Called when the request to the gRPC call (`callout_id`) fails or if the gRPC
+connection is closed.
+
+
+## Custom extension points
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_custom_callback`
```

Comment:
Is it possible to call back into Plugin context / Http Stream context / TCP stream context ?

How can extension understand for what context it is intended ?